community/rfcs/techniques/design/20200306-weight-clustering-design.md

```diff
@@ -0,0 +1,144 @@
+* [cluster_wrapper.py](../../tensorflow_model_optimization/python/core/clustering/keras/cluster_wrapper.py) and [cluster_wrapper_test.py](../../tensorflow_model_optimization/python/core/clustering/keras/cluster_wrapper_test.py) &mdash; implementation of a particular clustering algorithm and associated tests;
+* [clusterable_layer.py](../../tensorflow_model_optimization/python/core/clustering/keras/clusterable_layer.py) &mdash; similar to `PrunableLayer` &mdash; provides an abstract base class for custom layers to be recognized as clusterable;
+* [clustering_centroids.py](../../tensorflow_model_optimization/python/core/clustering/keras/clustering_centroids.py) and [clustering_centroids_test.py](../../tensorflow_model_optimization/python/core/clustering/keras/clustering_centroids_test.py) &mdash; implementations of different cluster centroid initialization techniques and tests for each of them;
+* [clustering_registry.py](../../tensorflow_model_optimization/python/core/clustering/keras/clustering_registry.py) and [clustering_registry_test.py](../../tensorflow_model_optimization/python/core/clustering/keras/clustering_registry_test.py) &mdash; similar to
+`PruningRegistry` &mdash; contains the list of clusterable parameters in each
+layer.
+
+## Clustering Workflow
+
+In general, the clustering workflow consists of several distinct steps, as follows:
+
+1. **Clustering** &mdash; run clustering on a trained model by invoking the `cluster_weights()` function with the desired clustering parameters:
+
+    ```python
+    clustered_model = cluster.cluster_weights(model, **clustering_params)
+    ```
+
+    NOTE: Clustering happens as part of the model building process, therefore in order to cluster sequential models, we need to call `model.build()` explicitly after applying `cluster_weights()` &mdash; see [the example](#example-0-cluster-an-entire-model) below.
+
+2. **Fine-tuning** &mdash; train the clustered model to fine-tune the initial values of cluster centroids:
+
+   ```python
+   clustered_model.fit(x_training_data, y_training_data, epochs=1)
+   ```
+
+3. **Stripping** &mdash; call `strip_clustering()` to strip the clustering-specific information from the model and re-obtain the original model structure:
+
+   ```python
+   stripped_model = cluster.strip_clustering(clustered_model)
+   ```
+
+The code snippets below illustrate common usage patterns of the clustering API. For the sake of brevity, the clustering examples presented here only deal with the 1st step of the above workflow.
+
+A more comprehensive end-to-end example on how to use the API to cluster and save a model is provided in the examples directory, under *clustering*.
+
+### Cluster an Entire Model
+
+When passing an entire model to `cluster_weights()`, the tool will determine which layers are clusterable and will cluster all of them &mdash; other layers will be ignored and left as they are.
```

Comment:
"other layers will be ignored and left as they are"

For pruning, the layers are still wrapped with the pruning wrapper. Is the proposal for it to be different for weight clustering?